presto-hive/src/main/java/com/facebook/presto/hive/auth/HadoopKerberosImpersonatingAuthentication.java

```diff
@@ -18,12 +18,34 @@
 package com.facebook.presto.hive.auth;
 
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.jetbrains.annotations.NotNull;
+import org.weakref.jmx.internal.guava.cache.CacheBuilder;
+import org.weakref.jmx.internal.guava.cache.CacheLoader;
+import org.weakref.jmx.internal.guava.cache.LoadingCache;
 
 public class HadoopKerberosImpersonatingAuthentication
         extends HadoopKerberosBaseAuthentication
 {
+    private LoadingCache<String, UserGroupInformation> userGroupInformationCache = CacheBuilder.newBuilder().build(
+            new CacheLoader<String, UserGroupInformation>()
+            {
+                @Override
+                public UserGroupInformation load(@NotNull String user)
+                        throws Exception
+                {
+                    return UserGroupInformation.createProxyUser(user, getUserGroupInformation());
+                }
+            });
+
     public HadoopKerberosImpersonatingAuthentication(String principal, String keytab, Configuration configuration)
     {
         super(principal, keytab, configuration);
     }
+
+    @Override
+    public UserGroupInformation getUserGroupInformation(String user)
+    {
+        return userGroupInformationCache.getUnchecked(user);
```

Comment:
perhaps you should call `checkTGTAndReloginFromKeytab` on the kerberos logged in `ugi` before returning the proxy user from cache